pkg/storage/watch_cache.go

```diff
@@ -250,6 +250,20 @@ func (w *watchCache) GetAllEventsSinceThreadUnsafe(resourceVersion uint64) ([]wa
 	if size > 0 {
 		oldest = w.cache[w.startIndex%w.capacity].resourceVersion
 	}
+	if resourceVersion == 0 {
+		// resourceVersion = 0 means that we don't require any specific starting point
+		// and we would like to start watching from ~now.
+		// However, to keep backward compatibility, we additionally need to return the
+		// current state and only then start watching from that point.
+		//
+		// TODO: In v2 api, we should stop returning the current state - #13969.
+		allItems := w.store.List()
```

Comment:
<
I say it's not "safe" because at no point in time can the client be certain it has a complete or correct list. I get that for many uses this safety isn't important.

I don't understand it btw. It is guaranteed that this list is a correct list from some point in time (as long as we don't have multi-object transactions, which is true for now).